### PR TITLE
Laboratory work #4, Ulyana Nagornaya - 22FPL2 

### PR DESCRIPTION
**Pull request by Ulyana-Nagornaya**

### CODE REVIEW

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 Top-p sampling generation and filling gaps with ngrams
 """
-# pylint:disable=too-few-public-methods, too-many-arguments
+from random import choice
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:6]: pleace, revert the pylint settings

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -28,6 +28,20 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(text, str) and text):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:32]: add message to the error

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -39,6 +53,11 @@ def _put(self, element: str) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(element, str) and element):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:57]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -56,7 +75,13 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(decoded_corpus, tuple) and decoded_corpus):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:79]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Add messages to each error

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -28,6 +32,20 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(text, str) and text):
+            raise ValueError ("WordProcessor._tokenize: inappropriate input")
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:36]: remove the prefix in the message. This information should be in traceback

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -39,6 +57,11 @@ def _put(self, element: str) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(element, str) and element):
+            raise ValueError ("WordProcessor._put: inappropriate input")
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:61]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -98,6 +130,40 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
                 or if sequence has inappropriate length,
                 or if methods used return None.
         """
+        if not (isinstance(prompt, str) and prompt and
+                isinstance(seq_len, int) and seq_len > 0):
+            raise ValueError ("TopPGenerator.run: inappropriate input")
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:135]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 8]
```diff
@@ -125,7 +194,8 @@ def get_conversion_generator_type(self, generator_type: int) -> str:  # type: ig
         Returns:
             (str): Name of the generator.
         """
-
+        generators = ['Greedy Generator', 'Top-P Generator', 'Beam Search Generator']
+        return generators[generator_type]
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:198]: what if  generator_type>3?

File: lab_4_fill_words_by_ngrams/main.py [hunk 9]
```diff
@@ -220,6 +301,31 @@ def _calculate_perplexity(self, generated_text: str) -> float:  # type: ignore
...
+            generated_tokens = self._language_model.generate_next_token(context)
+            if not generated_tokens:
+                raise ValueError ('QualityChecker._calculate_perplexity: none generated tokens')
+
+            probability = generated_tokens.get(encoded_seq[index])
+            if probability:
+                log_prob += log(probability)
+
+        if not log_prob:
+            raise ValueError ('QualityChecker._calculate_perplexity: sum of probabilities is 0')
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:325]: remove space

File: lab_4_fill_words_by_ngrams/main.py [hunk 10]
```diff
@@ -318,6 +466,11 @@ def __init__(
                 NGramLanguageModel instance to use for text generation
             word_processor (WordProcessor): WordProcessor instance to handle text processing
         """
+        self._generator_type = generator_type
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:469]: do you really need to save it?

File: lab_4_fill_words_by_ngrams/main.py [hunk 11]
```diff
@@ -343,3 +510,5 @@ def get_generator_type(self) -> str:  # type: ignore
         Returns:
             str: Generator type
         """
+        return GeneratorTypes().get_conversion_generator_type(self._generator_type)
+
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:514]: remove spaces

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 5 above)
Comment (Ulyana-Nagornaya) [reply lab_4_fill_words_by_ngrams/main.py:36]: OkAlrightThenGIF

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**